drivers/src/csrng.rs

```diff
@@ -0,0 +1,346 @@
+        Reseed = 3,
+        Generate = 4,
+        Update = 5,
+    }
+}
+
+// https://opentitan.org/book/hw/ip/csrng/doc/theory_of_operation.html#command-description
+const MAX_SEED_WORDS: usize = 12;
+const MAX_GENERATE_BLOCKS: usize = 4096;
+const WORDS_PER_GENERATE_BLOCK: usize = 4;
+
+/// A unique handle to the underlying CSRNG peripheral.
+pub struct Csrng {
+    // Force API consumers to use one of the construct methods to properly seed
+    // the CSRNG.
+    _prevent_struct_expression: (),
+}
+
+impl Csrng {
+    /// Returns a handle to the CSRNG in TRNG mode.
+    ///
+    /// The CSRNG will gather seed material from the entropy_src peripheral.
+    ///
+    /// # Safety
+    ///
+    /// No other handles to the CSRNG should exist.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the internal seed command fails.
+    pub unsafe fn new() -> CaliptraResult<Self> {
+        Self::with_seed(Seed::EntropySrc)
+    }
+
+    /// Returns a handle to the CSRNG configured to use the provided [`Seed`].
+    ///
+    /// # Safety
+    ///
+    /// No other handles to the CSRNG should exist.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the internal seed command fails.
+    pub unsafe fn with_seed(seed: Seed) -> CaliptraResult<Self> {
+        const FALSE: u32 = MultiBitBool::False as u32;
+        const TRUE: u32 = MultiBitBool::True as u32;
+
+        // Configure and enable entropy_src if needed.
+        let e = entropy_src::RegisterBlock::entropy_src_reg();
+
+        if e.module_enable().read().module_enable() == FALSE {
+            e.conf()
+                .write(|w| w.fips_enable(FALSE).entropy_data_reg_enable(FALSE));
+            e.module_enable().write(|w| w.module_enable(TRUE));
+            wait::until(|| e.debug_status().read().main_sm_boot_done());
+        }
+
+        if registers().ctrl().read().enable() == FALSE {
+            registers()
+                .ctrl()
+                .write(|w| w.enable(TRUE).sw_app_enable(TRUE).read_int_state(TRUE));
+        }
+
+        send_command(Command::Uninstantiate)?;
+        send_command(Command::Instantiate(seed))?;
+
+        Ok(Self {
+            _prevent_struct_expression: (),
+        })
+    }
+
+    /// Returns an iterator over `num_words` random [`u32`]s.
+    ///
+    /// This function will round up to the nearest multiple of four words.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the internal generate command fails.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// let mut csrng = ...;
+    ///
+    /// let num_words = NonZeroUsize::new(1).unwrap();
+    /// let mut random_words = csrng.generate(num_words)?;
+    ///
+    /// // Rounds up to nearest multiple of four.
+    /// assert_eq!(random_words.len(), 4);
+    ///
+    /// for word in random_words {
+    ///     // Do something with `word`.
+    /// }
+    /// ```
+    pub fn generate(&mut self, num_words: NonZeroUsize) -> CaliptraResult<Iter> {
```

Comment:
The normal use-case will be to initialize blinding vectors; I wonder if codegen might be tighter if this were just

```
pub fn generate(&mut self, buf: &[mut u32]);
```

Have you done any experiments?

Reply:
Played around with the CSRNG test binary and pretty much concluded that you can get any result you want depending on how you set up the call site, change the frequency of functions called, or pick which functions are `#[inline(always/never)]`.

@korran and I discussed offline that it's more useful to examine the code size of the driver once it's embedded in the ROM.

Below are a few experiments just to rule out anything egregious.

## Current PR
* Iterator fills a buffer.
* Buffer asserted against expected values.
```
$ cargo size --locked --target riscv32imc-unknown-none-elf --features emu,riscv --no-default-features --profile firmware -p caliptra-drivers-test-bin --bin csrng -- --format=sysv

section                size         addr
.text                  9149          0x0
.data                  1336   0x50000000
```

9149 + 1336 = 10485 bytes

## Point-of-use iterator
* Each value of iterator used immediately.
* No storage (beyond a single `u32`) of yielded values.
```
section                size         addr
.text                  9125          0x0
.data                  1256   0x50000000
```

9125 + 1256 = 10381 bytes

```rust
let mut csrng = unsafe { Csrng::with_seed(SEED) }.expect("construct CSRNG");
let num_words = NonZeroUsize::new(EXPECTED_OUTPUT.len()).expect("non-zero num_words");
let words = csrng.generate(num_words).expect("generate 2");

assert!(words.len() >= EXPECTED_OUTPUT.len());

for (actual, expected) in zip(words, EXPECTED_OUTPUT) {
    // Point-of-use of each random word as opposed to store-in-buffer.
    assert_eq!(actual, expected);
}
```

## Fill buffer
* Populate a caller-provided buffer.
* Buffer asserted against expected values.
```
section                size         addr
.text                  9053          0x0
.data                  1360   0x50000000
```

9053 + 1360 = 10413 bytes

```rust
pub fn generate(&mut self, out: &mut [u32]) -> CaliptraResult<()> {
    // Round up to nearest multiple of 128-bit block.
    let num_128_bit_blocks = (out.len() + 3) / 4;
    send_command(Command::Generate { num_128_bit_blocks })?;

    for (i, word) in out.iter_mut().enumerate() {
        if i % WORDS_PER_GENERATE_BLOCK == 0 {
            // Wait for CSRNG to generate next block of 4 words.
            wait::until(|| registers().genbits_vld().read().genbits_vld());
        }

        *word = registers().genbits().read();
    }

    Ok(())
}
```

Reply:
Ok, let's plan to go with the output buffer. That can be done in a separate PR if you like.

Reply:
Ok, let's defer this decision until we see what the ROM codegen looks like.